File: v4/rust/iprange-livedb/src/key.rs
//! Numeric IPv4 and IPv6 keys used by the exact v4 range map.
//!
//! Checked increment/decrement implement the numeric boundary rules used by
//! mutation code to trim at `from - 1` and `to + 1`.

use crate::cardinality::Cardinality129;
use crate::contract::AddressFamily;
use crate::error::{Error, Result};
use crate::mapping::ByteSource;

/// Common private interface over the two key widths, so physical algorithms are
/// written once and width-specialized at compile time.
pub(crate) trait IpKey: Copy + Ord + core::fmt::Debug + 'static {
    /// Key width in bytes (4 or 16).
    const WIDTH: usize;
    /// The IP family.
    const FAMILY: AddressFamily;

    /// Deserialize a key from the first [`WIDTH`](Self::WIDTH) bytes of `src`. Panics
    /// if `src` is shorter than `WIDTH`.
    fn read_le<S: ByteSource>(src: S, at: usize) -> Self;

    /// Serialize the key into the first [`WIDTH`](Self::WIDTH) bytes.
    fn write_le(self, output: &mut [u8]);

    fn checked_next(self) -> Option<Self>;
    fn checked_previous(self) -> Option<Self>;
    fn inclusive_cardinality(self, to: Self) -> Result<Cardinality129>;
}

/// An IPv4 address as a big-endian-valued `u32` (e.g. `192.0.2.1` = `0xC000_0201`),
/// stored little-endian on disk. Derived `Ord` is numeric.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
pub struct Ipv4Key(pub u32);

impl Ipv4Key {
    pub const MIN: Self = Self(0);
    pub const MAX: Self = Self(u32::MAX);

    #[inline]
    pub const fn checked_next(self) -> Option<Self> {
        match self.0.checked_add(1) {
            Some(value) => Some(Self(value)),
            None => None,
        }
    }

    #[inline]
    pub const fn checked_previous(self) -> Option<Self> {
        match self.0.checked_sub(1) {
            Some(value) => Some(Self(value)),
            None => None,
        }
    }
}

impl IpKey for Ipv4Key {
    const WIDTH: usize = 4;
    const FAMILY: AddressFamily = AddressFamily::Ipv4;

    #[inline]
    fn read_le<S: ByteSource>(src: S, at: usize) -> Self {
        Ipv4Key(u32::from_le_bytes(
            src.array(at).expect("validated IPv4 key"),
        ))
    }

    #[inline]
    fn write_le(self, output: &mut [u8]) {
        output[..4].copy_from_slice(&self.0.to_le_bytes());
    }

    #[inline]
    fn checked_next(self) -> Option<Self> {
        self.checked_next()
    }

    #[inline]
    fn checked_previous(self) -> Option<Self> {
        self.checked_previous()
    }

    #[inline]
    fn inclusive_cardinality(self, to: Self) -> Result<Cardinality129> {
        Cardinality129::ipv4_inclusive(self.0, to.0)
            .map_err(|_| Error::ArithmeticOverflow("IPv4 interval cardinality"))
    }
}

/// An IPv6 address as a `(hi, lo)` pair of `u64` — `hi` is the most-significant 64
/// bits. The struct keeps `(hi, lo)` so derived `Ord` is numeric; exact v4 wire
/// bytes are one little-endian `u128`, therefore the low limb is stored first.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
pub struct Ipv6Key {
    /// Most-significant 64 bits.
    pub hi: u64,
    /// Least-significant 64 bits.
    pub lo: u64,
}

impl Ipv6Key {
    pub const MIN: Self = Self { hi: 0, lo: 0 };
    pub const MAX: Self = Self {
        hi: u64::MAX,
        lo: u64::MAX,
    };

    /// Construct from the full 128-bit value.
    #[inline]
    pub const fn from_u128(v: u128) -> Self {
        Ipv6Key {
            hi: (v >> 64) as u64,
            lo: v as u64,
        }
    }

    /// The full 128-bit value.
    #[inline]
    pub const fn to_u128(self) -> u128 {
        ((self.hi as u128) << 64) | (self.lo as u128)
    }

    #[inline]
    pub const fn checked_next(self) -> Option<Self> {
        if self.hi == u64::MAX && self.lo == u64::MAX {
            None
        } else if self.lo == u64::MAX {
            Some(Self {
                hi: self.hi + 1,
                lo: 0,
            })
        } else {
            Some(Self {
                hi: self.hi,
                lo: self.lo + 1,
            })
        }
    }

    #[inline]
    pub const fn checked_previous(self) -> Option<Self> {
        if self.hi == 0 && self.lo == 0 {
            None
        } else if self.lo == 0 {
            Some(Self {
                hi: self.hi - 1,
                lo: u64::MAX,
            })
        } else {
            Some(Self {
                hi: self.hi,
                lo: self.lo - 1,
            })
        }
    }
}

impl IpKey for Ipv6Key {
    const WIDTH: usize = 16;
    const FAMILY: AddressFamily = AddressFamily::Ipv6;

    #[inline]
    fn read_le<S: ByteSource>(src: S, at: usize) -> Self {
        let l = src.array(at).expect("validated IPv6 low limb");
        let h = src.array(at + 8).expect("validated IPv6 high limb");
        Ipv6Key {
            hi: u64::from_le_bytes(h),
            lo: u64::from_le_bytes(l),
        }
    }

    #[inline]
    fn write_le(self, output: &mut [u8]) {
        output[..8].copy_from_slice(&self.lo.to_le_bytes());
        output[8..16].copy_from_slice(&self.hi.to_le_bytes());
    }

    #[inline]
    fn checked_next(self) -> Option<Self> {
        self.checked_next()
    }

    #[inline]
    fn checked_previous(self) -> Option<Self> {
        self.checked_previous()
    }

    #[inline]
    fn inclusive_cardinality(self, to: Self) -> Result<Cardinality129> {
        Cardinality129::ipv6_inclusive(self.hi, self.lo, to.hi, to.lo)
            .map_err(|_| Error::ArithmeticOverflow("IPv6 interval cardinality"))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn big_endian_portable_ipv6_key_matches_literal_bytes() {
        // 2001:db8::1 as one little-endian u128: low limb, then high limb.
        let k = Ipv6Key {
            hi: 0x2001_0db8_0000_0000,
            lo: 0x1,
        };
        let mut buf = [0u8; 16];
        k.write_le(&mut buf);
        let expected: [u8; 16] = [
            0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // lo, little-endian
            0x00, 0x00, 0x00, 0x00, 0xb8, 0x0d, 0x01, 0x20, // hi, little-endian
        ];
        assert_eq!(
            buf, expected,
            "IPv6 key on-disk bytes must match the encoding"
        );
        assert_eq!(Ipv6Key::read_le(&buf, 0), k, "round-trip");
        assert_eq!(k.to_u128(), 0x2001_0db8_0000_0000_0000_0000_0000_0001);
    }

    #[test]
    fn big_endian_portable_ipv4_key_matches_literal_bytes() {
        // 192.0.2.1 = 0xC000_0201 -> LE bytes 01 02 00 c0.
        let k = Ipv4Key(0xC000_0201);
        let mut buf = [0; 4];
        k.write_le(&mut buf);
        assert_eq!(buf, [0x01, 0x02, 0x00, 0xc0]);
        assert_eq!(Ipv4Key::read_le(&buf, 0), k);
    }

    #[test]
    fn ipv6_numeric_order_not_bytewise() {
        let a = Ipv6Key { hi: 1, lo: 0 };
        let b = Ipv6Key {
            hi: 0,
            lo: u64::MAX,
        };
        assert!(a > b, "compare hi then lo, not raw bytes");
        assert!(a.to_u128() > b.to_u128());
    }

    #[test]
    fn checked_inc_v6_carry_and_max() {
        assert_eq!(
            Ipv6Key {
                hi: 5,
                lo: u64::MAX
            }
            .checked_next(),
            Some(Ipv6Key { hi: 6, lo: 0 }),
            "carry from lo into hi"
        );
        assert_eq!(
            Ipv6Key { hi: 0, lo: 41 }.checked_next(),
            Some(Ipv6Key { hi: 0, lo: 42 })
        );
        assert_eq!(Ipv6Key::MAX.checked_next(), None, "no +1 at family_max");
    }

    #[test]
    fn checked_dec_v6_borrow_and_min() {
        assert_eq!(
            Ipv6Key { hi: 6, lo: 0 }.checked_previous(),
            Some(Ipv6Key {
                hi: 5,
                lo: u64::MAX
            }),
            "borrow from hi into lo"
        );
        assert_eq!(Ipv6Key::MIN.checked_previous(), None, "no -1 at family_min");
    }

    #[test]
    fn checked_inc_dec_v4_bounds() {
        assert_eq!(Ipv4Key(41).checked_next(), Some(Ipv4Key(42)));
        assert_eq!(Ipv4Key::MAX.checked_next(), None);
        assert_eq!(Ipv4Key(42).checked_previous(), Some(Ipv4Key(41)));
        assert_eq!(Ipv4Key::MIN.checked_previous(), None);
    }
}

File: v4/go/internal/exactv4/key.go
package exactv4

import (
	"encoding/binary"
	"math/bits"
)

type rangeKey[K any] interface {
	comparable
	writeLE([]byte)
	readLE([]byte) K
	fromHalves(uint64, uint64) K
	compare(K) int
	minimum() K
	maximum() K
	width() int
	family() AddressFamily
	halves() (uint64, uint64)
}

// IPv4 is one IPv4 address in numeric network-address order.
type IPv4 uint32

func (k IPv4) writeLE(dst []byte)                { binary.LittleEndian.PutUint32(dst, uint32(k)) }
func (IPv4) readLE(src []byte) IPv4              { return IPv4(binary.LittleEndian.Uint32(src)) }
func (IPv4) fromHalves(_ uint64, lo uint64) IPv4 { return IPv4(uint32(lo)) }
func (IPv4) minimum() IPv4                       { return 0 }
func (IPv4) maximum() IPv4                       { return ^IPv4(0) }
func (IPv4) width() int                          { return 4 }
func (IPv4) family() AddressFamily               { return AddressFamilyIPv4 }
func (k IPv4) halves() (uint64, uint64)          { return 0, uint64(k) }

func (k IPv4) compare(other IPv4) int {
	switch {
	case k < other:
		return -1
	case k > other:
		return 1
	default:
		return 0
	}
}

// Next returns the following IPv4 address, or false at the family maximum.
func (k IPv4) Next() (IPv4, bool) {
	if k == ^IPv4(0) {
		return 0, false
	}
	return k + 1, true
}

// Previous returns the preceding IPv4 address, or false at the family minimum.
func (k IPv4) Previous() (IPv4, bool) {
	if k == 0 {
		return 0, false
	}
	return k - 1, true
}

// IPv6 is one IPv6 address as numeric high and low 64-bit halves.
type IPv6 struct {
	Hi uint64
	Lo uint64
}

// IPv6FromHalves constructs an IPv6 address from its numeric halves.
func IPv6FromHalves(hi, lo uint64) IPv6 { return IPv6{Hi: hi, Lo: lo} }

// Exact v4 stores one little-endian u128, so the low limb comes first.
func (k IPv6) writeLE(dst []byte) {
	binary.LittleEndian.PutUint64(dst[0:8], k.Lo)
	binary.LittleEndian.PutUint64(dst[8:16], k.Hi)
}

func (IPv6) readLE(src []byte) IPv6 {
	return IPv6{
		Hi: binary.LittleEndian.Uint64(src[8:16]),
		Lo: binary.LittleEndian.Uint64(src[0:8]),
	}
}

func (IPv6) fromHalves(hi, lo uint64) IPv6 { return IPv6{Hi: hi, Lo: lo} }

func (IPv6) minimum() IPv6              { return IPv6{} }
func (IPv6) maximum() IPv6              { return IPv6{Hi: ^uint64(0), Lo: ^uint64(0)} }
func (IPv6) width() int                 { return 16 }
func (IPv6) family() AddressFamily      { return AddressFamilyIPv6 }
func (k IPv6) halves() (uint64, uint64) { return k.Hi, k.Lo }

func (k IPv6) compare(other IPv6) int {
	if k.Hi < other.Hi {
		return -1
	}
	if k.Hi > other.Hi {
		return 1
	}
	if k.Lo < other.Lo {
		return -1
	}
	if k.Lo > other.Lo {
		return 1
	}
	return 0
}

// Next returns the following IPv6 address, or false at the family maximum.
func (k IPv6) Next() (IPv6, bool) {
	if k == (IPv6{Hi: ^uint64(0), Lo: ^uint64(0)}) {
		return IPv6{}, false
	}
	lo, carry := bits.Add64(k.Lo, 1, 0)
	return IPv6{Hi: k.Hi + carry, Lo: lo}, true
}

// Previous returns the preceding IPv6 address, or false at the family minimum.
func (k IPv6) Previous() (IPv6, bool) {
	if k == (IPv6{}) {
		return IPv6{}, false
	}
	lo, borrow := bits.Sub64(k.Lo, 1, 0)
	return IPv6{Hi: k.Hi - borrow, Lo: lo}, true
}
